modules/patching/diff.go

```diff
@@ -0,0 +1,275 @@
+		for _, current := range intermediateDiffs {
+			switch {
+			// CASE 1: IndexA < IndexB
+			case other.StartIndex < current.StartIndex:
+				switch {
+				// CASE 1a, 1b: Ins - Ins, Ins - Rmv
+				case other.Insertion && current.Insertion, other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType2, newIntermediateDiffs, current, other)
+				// CASE 1c: Rmv - Ins
+				case !other.Insertion && current.Insertion:
+					newIntermediateDiffs = doTransform(transformType3, newIntermediateDiffs, current, other)
+				// CASE 1d: Rmv - Rmv
+				case !other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType4, newIntermediateDiffs, current, other)
+				// Fail; should never have gotten to here.
+				default:
+					panic(fmt.Sprintf("Got to invalid state 1e while transforming [%s] on predessor [%+v], from list [%+v]", current.String(), other, others))
+				}
+			// CASE 2: IndexA = IndexB
+			case other.StartIndex == current.StartIndex:
+				switch {
+				// CASES 2a, 2b: Ins - Ins, Ins - Rmv
+				case other.Insertion && current.Insertion, other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType2, newIntermediateDiffs, current, other)
+				// CASE 2c: Rmv - Ins
+				// Do nothing
+				case !other.Insertion && current.Insertion:
+					newIntermediateDiffs = doTransform(transformType1, newIntermediateDiffs, current, other)
+				// CASE 2d: Rmv - Ins
+				case !other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType5, newIntermediateDiffs, current, other)
+				// Fail; should never have gotten to here.
+				default:
+					panic(fmt.Sprintf("Got to invalid state 2e while transforming [%s] on predessor [%+v], from list [%+v]", current.String(), other, others))
+				}
+			// CASE 3: IndexA > IndexB
+			case other.StartIndex > current.StartIndex:
+				switch {
+				// CASE 3a, 3c: Ins - Ins, Rmv - Ins
+				case other.Insertion && current.Insertion, !other.Insertion && current.Insertion:
+					newIntermediateDiffs = doTransform(transformType1, newIntermediateDiffs, current, other)
+				// CASE 3b: Ins - Rmv
+				case other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType6, newIntermediateDiffs, current, other)
+				// CASE 3d: Rmv - Rmv
+				case !other.Insertion && !current.Insertion:
+					newIntermediateDiffs = doTransform(transformType7, newIntermediateDiffs, current, other)
+				// Fail; should never have gotten to here.
+				default:
+					panic(fmt.Sprintf("Got to invalid state 3e while transforming [%s] on predessor [%+v], from list [%+v]", current.String(), other, others))
+				}
+			}
+		}
+		intermediateDiffs = newIntermediateDiffs
+	}
+	return intermediateDiffs
+}
+
+func doTransform(transformFunc func(current, other *Diff) []*Diff, currResults []*Diff, current, other *Diff) []*Diff {
+	return append(currResults, transformFunc(current, other)...)
```

Comment:
Append uses the same algorithm as ArrayLists in java - they're Amortized O(1) by doubling the capacity of the slice. It's actually not that bad.